### PR TITLE
journalctl: split journalctl.c into small pieces

### PR DESCRIPTION
**Pull request by yuwata**
Almost no functional change, just refactoring.

### CODE REVIEW

File: src/journal/journalctl-misc.c [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+
+        printf("Archived and active journals take up %s in the file system.\n", FORMAT_BYTES(bytes));
+        return 0;
+}
+
+int action_list_field_names(void) {
+        _cleanup_(sd_journal_closep) sd_journal *j = NULL;
+        int r;
+
+        assert(arg_action == ACTION_LIST_FIELDS);
```
Comment (yuwata) [inline src/journal/journalctl-misc.c]: TODO: fix this line.

File: src/journal/journalctl-authenticate.c [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+        r = stat("/var/log/journal", &st);
+        if (r < 0 && !IN_SET(errno, ENOENT, ENOTDIR))
+                return log_error_errno(errno, "stat(\"%s\") failed: %m", "/var/log/journal");
+
+        if (r < 0 || !S_ISDIR(st.st_mode)) {
+                log_error("%s is not a directory, must be using persistent logging for FSS.",
+                          "/var/log/journal");
+                return r < 0 ? -errno : -ENOTDIR;
+        }
```
Comment (YHNdnzj) [inline src/journal/journalctl-authenticate.c:80]: Should probably add `verify_directory_at` and use that here?

File: src/journal/meson.build [hunk 3]
```diff
@@ -29,6 +29,20 @@ libjournal_core = static_library(
...
+        'journalctl-catalog.c',
+        'journalctl-filter.c',
+        'journalctl-misc.c',
+        'journalctl-show.c',
+        'journalctl-util.c',
+        'journalctl-varlink.c',
+)
+
+if conf.get('HAVE_GCRYPT') == 1
+        journalctl_sources += files('journalctl-authenticate.c',)
```
Comment (YHNdnzj) [inline src/journal/meson.build]:
```suggestion
        journalctl_sources += files('journalctl-authenticate.c')
```

File: src/journal/journalctl.h [hunk 4]
```diff
@@ -0,0 +1,99 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+#pragma once
+
+#include <stdbool.h>
+#include <inttypes.h>
```
Comment (YHNdnzj) [inline src/journal/journalctl.h]: sort alphabetically?

File: src/journal/journalctl-util.c [hunk 5]
```diff
@@ -0,0 +1,69 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "journal-util.h"
+#include "journalctl.h"
+#include "journalctl-util.h"
+#include "rlimit-util.h"
+#include "sigbus.h"
+#include "terminal-util.h"
+
+char* format_timestamp_maybe_utc(char *buf, size_t l, usec_t t) {
```
Comment (YHNdnzj) [inline src/journal/journalctl-util.c:10]: Add assertions?

File: src/journal/journalctl-util.c [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        _cleanup_(sd_journal_closep) sd_journal *j = NULL;
+        int r;
+
+        assert(ret);
+
+        /* Increase max number of open files if we can, we might needs this when browsing journal files, which might be
+         * split up into many files. */
+        (void) rlimit_nofile_bump(HIGH_RLIMIT_NOFILE);
+
+        signal(SIGWINCH, columns_lines_cache_reset);
```
Comment (YHNdnzj) [inline src/journal/journalctl-util.c]: Hmm, this should be moved to action_show?

File: src/journal/journalctl-util.c [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        sigbus_install();
```
Comment (YHNdnzj) [inline src/journal/journalctl-util.c:29]: This should stay in run()?

File: src/journal/journalctl-show.c [hunk 8]
```diff
@@ -0,0 +1,478 @@
...
+        int n_shown, r, poll_fd = -EBADF;
+
+        assert(arg_action == ACTION_SHOW);
+
+        r = acquire_journal(&c.journal);
+        if (r < 0)
+                return r;
+
+        if (!journal_has_fields(c.journal)) {
+                if (arg_action == ACTION_SHOW && arg_compiled_pattern)
```
Comment (YHNdnzj) [inline src/journal/journalctl-show.c]: The first condition is now redundant?

File: src/journal/journalctl-util.c [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        r = journal_access_check_and_warn(j, arg_quiet,
+                                          !(arg_journal_type == SD_JOURNAL_CURRENT_USER || arg_user_units));
+        if (r < 0)
+                return r;
+
+        *ret = TAKE_PTR(j);
+        return 0;
+}
+
+bool journal_has_fields(sd_journal *j) {
```
Comment (YHNdnzj) [inline src/journal/journalctl-util.c]: Hmm, should probably be called `journal_boot_has_effect` or sth? The current name is a bit confusing.

File: src/journal/journalctl-util.c (hunk 7 above)
Comment (yuwata) [reply src/journal/journalctl-util.c:29]: sigbus is only triggered when using sd-journal, so I'd like to keep this here.

File: src/journal/journalctl-util.c (hunk 9 above)
Comment (yuwata) [reply src/journal/journalctl-util.c]: Sounds nice. I have no idea why I named this so...

File: src/journal/journalctl-authenticate.c (hunk 2 above)
Comment (yuwata) [reply src/journal/journalctl-authenticate.c:80]: `is_dir()` is OK?

**Comment by yuwata** [comment]
@YHNdnzj Thank you for the detailed review. Most requests are addressed. Please see the exceptions. PTAL.

**Comment by yuwata** [comment]
@YHNdnzj I noticed you tagged this to v257, rather than v256. Do you want to postpone this?

**Comment by YHNdnzj** [comment]
> @YHNdnzj I noticed you tagged this to v257, rather than v256. Do you want to postpone this?

Not really :) I just wanted to give it a milestone, but don't want that to carry a meaning that this must go in very quickly (since we're late in the v256 release cycle). It's perfectly fine to merge it now if ready.

I'll finish my review tomorrow.

File: src/journal/journalctl-varlink.c [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+#include "errno-util.h"
+#include "journal-internal.h"
+#include "journal-vacuum.h"
+#include "journalctl.h"
+#include "journalctl-util.h"
+#include "journalctl-varlink.h"
+#include "varlink.h"
+
+static int varlink_connect_journal(Varlink **ret_link) {
```
Comment (YHNdnzj) [inline src/journal/journalctl-varlink.c]: Add assertion while at it?

File: src/journal/journalctl-authenticate.c [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+
+        assert(arg_interval > 0);
+
+        n = now(CLOCK_REALTIME);
+        n /= arg_interval;
+
+        safe_close(fd);
+        fd = mkostemp_safe(k);
+        if (fd < 0)
+                return log_error_errno(fd, "Failed to open %s: %m", k);
```
Comment (YHNdnzj) [inline src/journal/journalctl-authenticate.c:134]: Ugh, `fd` is definitely invalid here, so no need for `safe_close`.

One day we should probably switch this to `open_tmpfile_linkable`, but don't bother with that in this PR.

File: src/journal/journalctl-catalog.c [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+        assert(arg_action == ACTION_UPDATE_CATALOG);
+
+        database = path_join(arg_root, secure_getenv("SYSTEMD_CATALOG") ?: CATALOG_DATABASE);
+        if (!database)
+                return log_oom();
+
+        e = secure_getenv("SYSTEMD_CATALOG_SOURCES");
+        r = catalog_update(database,
+                           arg_root,
+                           e ? (const char* const*) STRV_MAKE(e) : catalog_file_dirs);
```
Comment (YHNdnzj) [inline src/journal/journalctl-catalog.c]: STRV_MAKE_CONST

File: src/journal/journalctl-catalog.c [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+        if (r < 0)
+                return log_error_errno(r, "Failed to list catalog: %m");
```
Comment (YHNdnzj) [inline src/journal/journalctl-catalog.c]:
```suggestion
                return log_error_errno(r, "Failed to update catalog: %m");
```

File: src/journal/journalctl-show.c [hunk 14]
```diff
@@ -0,0 +1,476 @@
...
+        bool since_seeked;
+        bool ellipsized;
+        bool previous_boot_id_valid;
+        sd_id128_t previous_boot_id;
+        sd_id128_t previous_boot_id_output;
+        dual_timestamp previous_ts_output;
+} Context;
+
+static void context_done(Context *c) {
+        if (!c)
```
Comment (YHNdnzj) [inline src/journal/journalctl-show.c]: This is stack-allocated, so should be assertion?

File: src/journal/journalctl-filter.c [hunk 15]
```diff
@@ -0,0 +1,543 @@
...
+        assert(j);
+
+        if (arg_priorities == 0xFF)
+                return 0;
+
+        for (i = LOG_EMERG; i <= LOG_DEBUG; i++)
+                if (arg_priorities & (1 << i)) {
+                        match[sizeof(match)-2] = '0' + i;
+
+                        r = sd_journal_add_match(j, match, strlen(match));
```
Comment (YHNdnzj) [inline src/journal/journalctl-filter.c:312]: `journal_add_matchf`?

File: src/journal/journalctl-filter.c [hunk 16]
```diff
@@ -0,0 +1,543 @@
...
+static int add_facilities(sd_journal *j) {
+        void *p;
+        int r;
+
+        SET_FOREACH(p, arg_facilities) {
+                char match[STRLEN("SYSLOG_FACILITY=") + DECIMAL_STR_MAX(int)];
+
+                xsprintf(match, "SYSLOG_FACILITY=%d", PTR_TO_INT(p));
+
+                r = sd_journal_add_match(j, match, strlen(match));
```
Comment (YHNdnzj) [inline src/journal/journalctl-filter.c:333]: also here

File: src/journal/journalctl-filter.c [hunk 17]
```diff
@@ -0,0 +1,543 @@
...
+                        return log_error_errno(r, "Failed to add match: %m");
+
+                if (sd_device_get_devname(d, &devnode) >= 0) {
+                        _cleanup_free_ char *match1 = NULL;
+
+                        r = stat(devnode, &st);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to stat() device node \"%s\": %m", devnode);
+
+                        r = asprintf(&match1, "_KERNEL_DEVICE=%c" DEVNUM_FORMAT_STR, S_ISBLK(st.st_mode) ? 'b' : 'c', DEVNUM_FORMAT_VAL(st.st_rdev));
```
Comment (YHNdnzj) [inline src/journal/journalctl-filter.c:389]: and here

File: src/journal/journalctl-filter.c [hunk 18]
```diff
@@ -0,0 +1,543 @@
...
+
+static int add_syslog_identifier(sd_journal *j) {
+        int r;
+
+        assert(j);
+
+        STRV_FOREACH(i, arg_syslog_identifier) {
+                _cleanup_free_ char *u = NULL;
+
+                u = strjoin("SYSLOG_IDENTIFIER=", *i);
```
Comment (YHNdnzj) [inline src/journal/journalctl-filter.c:268]: journal_add_match_pair

**Review by YHNdnzj — APPROVED** [review]

**Comment by yuwata** [comment]
@YHNdnzj Thank you for the review. Half of them are already waiting in another PR, or an existing issue. I will address them in a later PR.

**Pull request merged**